lib/charms/vault_k8s/v0/juju_facade.py

```diff
@@ -331,97 +377,157 @@ def get_relation_by_id(self, relation_name: str, relation_id: int) -> Relation:
             raise NoSuchRelationError(f"Relation {relation_name}:{relation_id} not found")
         return relation
 
-    def get_relations(self, relation_name: str) -> List[Relation]:
+    def get_relation_by_name(self, relation_name: str) -> Relation:
+        """Get the relation object by name.
+
+        Raises:
+            NoSuchRelationError
+            MultipleRelationsFoundError
+        """
+        relations = self.charm.model.relations.get(relation_name, [])
+        if not relations:
+            logger.error("No relations found for %s", relation_name)
+            raise NoSuchRelationError(f"Relation {relation_name} not found")
+        if len(relations) > 1:
+            logger.warning("More than one relation found for %s", relation_name)
+            raise MultipleRelationsFoundError(
+                f"More than one relation found for name {relation_name}"
+            )
+        return relations[0]
+
+    def get_relations(self, relation_name: str, relation_id: int | None = None) -> List[Relation]:
         """Get all relation objects with the given name.
 
         Returns:
             A list of relation objects, the list is empty if no relations are found
         """
-        relations = self.charm.model.relations.get(relation_name, [])
+        relations = (
+            self.charm.model.relations.get(relation_name, [])
+            if not relation_id
+            else [self.get_relation_by_id(relation_name, relation_id)]
+        )
         if not relations:
             logger.error("No relations found for %s", relation_name)
         return relations
 
+    def get_active_relations(
+        self, relation_name: str, relation_id: int | None = None
+    ) -> List[Relation]:
+        """Get all relations with the given name or ID that are active."""
+        relations = self.get_relations(relation_name, relation_id)
+        return [relation for relation in relations if relation.active]
+
     def relation_exists(self, relation_name: str) -> bool:
         """Check if there are any relations with the given name."""
         return self.get_relations(relation_name) != []
 
     def _read_relation_data(
-        self, relation_name: str, relation_id: int, entity: Unit | Application
+        self, relation_name: str, relation_id: int | None, entity: Unit | Application
     ) -> RelationDataContent | dict[str, str]:
-        relation = self.get_relation_by_id(relation_name, relation_id)
-        if not relation:
-            raise NoSuchRelationError(f"Relation {relation_name}:{relation_id} not found")
+        try:
+            relation = (
+                self.get_relation_by_id(relation_name, relation_id)
+                if relation_id
+                else self.get_relation_by_name(relation_name)
+            )
+        except (NoSuchRelationError, MultipleRelationsFoundError):
+            raise
```

Comment:
Similar comment, what does this try/except block do?

Reply:
We have those block all over the place, but we simply re-raise the same error, why catch it at all then?

Reply:
Removed them all. Added the exception to the docstring in public functions where it is needed to signal to the caller the possible exceptions.
More code reduction btw :)